### PR TITLE
INFRA-6555 - add support for password_reset_required=True to AWS user creation

### PR DESCRIPTION
**Pull request by tkwilliams**
- Add new option `password_reset_required` (defaults to False) to `boto_iam.create_login_profile()`
- Add new function `boto_iam.update_login_profile()`
- Expose `password_reset_required` to state function `boto_iam.user_present()`

### CODE REVIEW

**Comment by tkwilliams** [comment]
Can't see that any of the failing tests have anything to do with my changes... :)

File: salt/modules/boto_iam.py [hunk 1]
```diff
@@ -740,8 +740,34 @@ def delete_group(group_name, region=None, key=None,
...
+    except botocore.exceptions.ClientError as e:
+        log.debug(e)
+        err = e.response.get('Error', {}).get('Message')
+        # Error.Code is 'NoSuchEntity' for both a missing user and a missing Login Profile...
+        if 'Cannot find Login Profile for User' not in err:
+            log.error('Failed to get Login Profile for IAM user {0}: {1}'.format(user_name, err))
+        return None
+
+
+def create_login_profile(user_name, password, password_reset_required=False,
```
Comment (rallytime) [inline salt/modules/boto_iam.py]: Addubg `password_reset_required` in the beginning of the kwargs list will cause an  API change. Can you move that to the end of the list please?

File: salt/modules/boto_iam.py [hunk 2]
```diff
@@ -740,8 +740,34 @@ def delete_group(group_name, region=None, key=None,
...
 
 
-def create_login_profile(user_name, password, region=None, key=None,
-                         keyid=None, profile=None):
+def get_login_profile(user_name, region=None, key=None, keyid=None, profile=None):
+    '''
+    Return current Login Profile attached to user, if the user exists and
+    has a profile attached, None otherwise.
+
+    CLI Example:
```
Comment (rallytime) [inline salt/modules/boto_iam.py]: Can you add a `.. versionadded:: Fluorine` tag to this new function?

Each arg/kwarg needs documentation as well here. Can you add that?

File: salt/modules/boto_iam.py [hunk 3]
```diff
@@ -754,21 +780,48 @@ def create_login_profile(user_name, password, region=None, key=None,
...
-    conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)
+
+
+def update_login_profile(user_name, password=None, password_reset_required=False,
+                         region=None, key=None, keyid=None, profile=None):
+    '''
+    Updates a login profile for the specified user, give the user the
+    ability to access AWS services and the AWS Management Console.
+
+    .. versionadded:: 2015.8.0
```
Comment (rallytime) [inline salt/modules/boto_iam.py]: This should be `Fluorine`.

Comment (rallytime) [reply salt/modules/boto_iam.py]: Can you also add documentation for each of these kwargs?

File: salt/states/boto_iam.py [hunk 4]
```diff
@@ -478,8 +478,8 @@ def _delete_key(ret, access_key_id, user_name, region=None, key=None, keyid=None
         return ret
 
 
-def user_present(name, policies=None, policies_from_pillars=None, managed_policies=None, password=None, path=None,
-                 region=None, key=None, keyid=None, profile=None):
+def user_present(name, policies=None, policies_from_pillars=None, managed_policies=None, password=None,
```
Comment (rallytime) [inline salt/states/boto_iam.py]: Why is the `path` option being removed here? What is the strategy for people using/expecting this argument to work?

**Review by rallytime — CHANGES_REQUESTED** [review]
Hi @tkwilliams! Thanks for adding this. I have some comments/requests on this. In addition, the new functions you've written need some tests.

**Comment by rallytime** [comment]
@tkwilliams yeah, those test failures are not related to you changes. We're working on cleaning those up. :)

File: salt/modules/boto_iam.py (hunk 1 above)
Comment (tkwilliams) [reply salt/modules/boto_iam.py]: I actually tracked down every usage of create_login_profile() extant in the code-base (there was only one - in the boto_iam state module) and updated to match this param ordering.

I can see your point vis-a-vis hand-jammed module function calls though.  I hate the inelegance of shoving a new password related param on the end, far away from `password` itself but I see where you're coming from.

File: salt/modules/boto_iam.py (hunk 3 above)
Comment (tkwilliams) [reply salt/modules/boto_iam.py]: That wasn't my `versionadded` tag - it's preexisting :)
Hrrm, I thought I DID add docu for my new args... looking....

File: salt/states/boto_iam.py (hunk 4 above)
Comment (tkwilliams) [reply salt/states/boto_iam.py]: Almost entirely certain that was a stupid typo.  Again looking :)

**Comment by tkwilliams** [comment]
@rallytime  BTW, is it OK if I re-tool this to be based against 2018.3 rather than develop?  I wasn't thinking when I PR'ed it and having it against 2018.3 would make life much simpler on my side.

Do I need to close this and create a new PR or is there a way to convert this one mebbe?  :)

File: salt/states/boto_iam.py (hunk 4 above)
Comment (tkwilliams) [reply salt/states/boto_iam.py]: Oh, it wasn't removed - it merely folded to the next line, which doesn't appear in the snippet above.  It IS still in there however :)

**Comment by tkwilliams** [comment]
OK, I think I've worked in all requested changes.  I've also rebuilt it (via cherry-picks and a bit of fiddlery) to be branched off of 2018.3 instead.  Hope it shows up that way when I push....

**Comment by tkwilliams** [comment]
@rallytime Doesn't appear to have picked up the re-branch-from-scratch against 2018.3 ...  Any magic I can apply or should I just close and re-cut the PR entirely?

**Comment by rallytime** [comment]
@tkwilliams 2018.3 is feature-frozen, so new functions will need to stay in develop I'm afraid.

To answer your question about "magic" to apply for the future, you would need to change the base branch you're submitting your PR against. (See the `EDIT` button near the title of your PR.)

**Comment by rallytime** [comment]
@tkwilliams This needs to be submitted against develop and needs some unit tests written against the new functions. I'm going to close this for now since we haven't heard back for a bit here. Please feel free to resubmit this against develop. Thanks again! :D

**Pull request closed without merging**